plugins/inputs/tencentcloudcm/tencentcloudcm.go

```diff
@@ -0,0 +1,371 @@
+						t.Accounts[i].Namespaces[j].Name)
+				}
+				t.Accounts[i].Namespaces[j].Metrics = metrics
+			}
+		}
+	}
+
+	t.client = &cloudmonitorClient{Accounts: t.Accounts}
+
+	return nil
+
+}
+
+func (t *TencentCloudCM) updateWindow(relativeTo time.Time) {
+
+	windowEnd := relativeTo.Add(-time.Duration(t.Delay))
+
+	// starting point is two times the aggregation period to make sure all points are covered
+	t.windowStart = windowEnd.Add(-time.Duration(t.Period) * 2)
+
+	t.windowEnd = windowEnd
+}
+
+func (t *TencentCloudCM) Gather(acc telegraf.Accumulator) error {
+	t.updateWindow(time.Now())
+
+	metricObjects := t.client.GetMetricObjects(*t)
+
+	lmtr := limiter.NewRateLimiter(t.RateLimit, time.Second)
+	defer lmtr.Stop()
+
+	wg := sync.WaitGroup{}
+	rLock := sync.Mutex{}
+	results := []monitor.GetMonitorDataResponse{}
+
+	// requestIDMap contains request ID and metric objects for later aggregation
+	requestIDMap := map[string]metricObject{}
+	for _, obj := range metricObjects {
+
+		wg.Add(1)
+		<-lmtr.C
+		go func(m metricObject) {
+			defer wg.Done()
+
+			for {
+
+				client, err := t.client.NewClient(m.Region, m.Account.crs, *t)
+				if err != nil {
+					acc.AddError(err)
+					break
+				}
+
+				size := 100
+				if len(m.MonitorInstances) >= size {
+					batch := m.MonitorInstances[:size]
+					if len(batch) == 0 {
+						break
+					}
+
+					request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, batch, *t)
+
+					result, err := t.client.GatherMetrics(client, request, *t)
+					if err != nil {
+						acc.AddError(err)
+						break
+					}
+
+					rLock.Lock()
+					requestIDMap[*result.Response.RequestId] = m
+					results = append(results, *result)
+					rLock.Unlock()
+
+					m.MonitorInstances = m.MonitorInstances[size:]
+				} else {
+					request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, m.MonitorInstances, *t)
+
+					result, err := t.client.GatherMetrics(client, request, *t)
+					if err != nil {
+						acc.AddError(err)
+						break
+					}
+
+					rLock.Lock()
+					requestIDMap[*result.Response.RequestId] = m
+					results = append(results, *result)
+					rLock.Unlock()
+					break
+				}
+
+			}
+
+		}(obj)
+
+	}
+	wg.Wait()
+	for _, result := range results {
+		for _, datapoints := range result.Response.DataPoints {
+			tags := map[string]string{}
+
+			keys := []string{}
+			for _, v := range datapoints.Dimensions {
+				tags[*v.Name] = *v.Value
+				keys = append(keys, *v.Value)
+			}
+
+			metricObject := requestIDMap[*result.Response.RequestId]
+
+			if metricObject.isDiscovered {
+				instance := t.discoverTool.GetInstance(metricObject.Account.Name, metricObject.Namespace, metricObject.Region, newKey(strings.Join(keys, "-")))
+				for k, v := range instance {
+					if v != nil && !strings.Contains(strings.ToLower(k), "time") && fmt.Sprintf("%v", v) != "" {
+						tags[fmt.Sprintf("_%s_%s", metricObject.Namespace, k)] = fmt.Sprintf("%v", v)
+					}
```

Comment:
The handling of the value `v` is _strange_. What type do you expect for `v`?

Reply:
`v` is supposed to be an interface that is `non-nil` with `non-empty` value.

Reply:
TBH, it's not very satisfying for me to have no idea what `v` is (usually). I see that it might differ from product to product, but why don't you do the conversion to string in that product. For each product you likely know what to expect and can do some proper and robust conversion...

Reply:
fixed in e250797.

The `v` is now converted and filtered inside `product.go`.

Reply:
Perfect thank you!